### PR TITLE
[NEW] HeatMap Widget for One2many and Many2many records

### PR DESCRIPTION
**Pull request by GabbasovDinar**
This module allows you to display the One2many and Many2many records as a calendar of heatmaps.

Example::

    <record id="view_partner_form" model="ir.ui.view">
        <field name="name">res.partner.view.buttons</field>
        <field name="model">res.partner</field>
        <field name="inherit_id" ref="base.view_partner_form" />
        <field name="priority" eval="20"/>
        <field name="groups_id" eval="[(4, ref('sales_team.group_sale_salesman'))]"/>
        <field name="arch" type="xml">
            <page name="sales_purchases" position="after">
                <page name='sale_orders' string="Orders">
                    <field name="sale_order_ids" widget="heatmap"/>
                </page>
            </page>
        </field>
    </record>

Result:

screenshot-misc odoo11 local-2019 06 10-17-53-40

Discussion: https://github.com/OCA/web/issues/1157

CC: @pedrobaeza @agenterp 
@it-projects-llc

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I'm wondering if this is worth more as view than as widget, as loading it as widget overloads the destination record a lot.

**Comment by agent-z28** [comment]
The initial idea is to show directly how to the order behaviour is. 
From performance site we'll have to check.

**Comment by pedrobaeza** [comment]
I agree on the feature, but Odoo paradigm is to put an smart-button on the partner (like the existing one with the sales count), and when clicking navigate to the other view (right now, to the sales order list, but you can add the same smart-button for navigating to the heat-map). Take into account than adding this heatmap at partner level, overloads the server having to load all sales order each time you access the partner, even for modifying other things like the name, query the email, etc.

**Comment by ilmir-k** [comment]
@pedrobaeza Thanks for your feedback. Actually we thought about making it as a view rather than widget but stayed on the latter at this moment. We'll take it into account.
cc @GabbasovDinar

**Comment by GabbasovDinar** [comment]
What is the difference between displaying records as a tree view or using this method? Both methods overload the server to load the entire sales order.

**Comment by pedrobaeza** [comment]
Exactly, and that's the reason why Odoo doesn't do neither of both inside res.partner. It always relies in an smart-button that navigates to the object.

**Comment by agent-z28** [comment]
Github is able to load the whole! page within 1 second, and shows the widget in the mainscreen.
I assume they generate the widget html in the background, like a depends method on commit. 
What about a similar solution?

**Comment by pedrobaeza** [comment]
Uhm, any cached solution always have the problem of out of sync data. Any way, the problem here is not the time it takes to render (which can be more or less depending on the volume data, the number of customizations in the target model, etc), the question is that we are going to render it uselessly a lot of times when accessing the partner for other reasons.

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+            '/web_widget_heatmap/static/lib/js/cal-heatmap.js'
+        ],
+
+        /**
+         * Calculate and get the maximum date (up to which date the grid will be built).
+         * Display of control buttons depends on the maximum date.
+         * @param {Object} options
+         * @returns {Date} max_date
+         */
+        get_max_date: function(options) {
```
Comment (tarteo) [inline web_widget_heatmap/static/src/js/heatmap_widget.js:44]: please use dromedary case.

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+    var AbstractField = require('web.AbstractField');
+    var registry = require('web.field_registry');
+
+
+    var HeatMapWidget = AbstractField.extend({
+
+        fieldsToFetch: {
+            display_name: {type: 'char'},
+            create_date: {type: 'datetime'},
```
Comment (tarteo) [inline web_widget_heatmap/static/src/js/heatmap_widget.js:17]: Can you make the date field configurable?

File: web_widget_heatmap/static/src/css/heatmap_widget.css [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/* Copyright 2019 Dinar Gabbasov <https://it-projects.info/team/GabbasovDinar>
+ * License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). */
+.heatmap_widget {
+	width: 100%;
```
Comment (tarteo) [inline web_widget_heatmap/static/src/css/heatmap_widget.css:4]: Can you use spaces here instead of tabs

**Review by tarteo — CHANGES_REQUESTED** [review]
I like this widget.
The loading problem can also be solved by making the widget work for Serialized fields. The programmer can make a computed and stored Serialized field which stores the records as JSON. This way every time the partner record opens, not all the sale orders are fetched instead just a text field essentially.

File: web_widget_heatmap/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2019 Dinar Gabbasov <https://it-projects.info/team/GabbasovDinar>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Web Widget HeatMap",
+    "summary": "Displaying your records in calendar HeatMap",
+    "version": "11.0.1.0.0",
+    "development_status": "Stable",
+    "category": "Extra Tools",
+    "website": "https://github.com/OCA/web/tree/11.0/web_widget_heatmap",
```
Comment (Tardo) [inline web_widget_heatmap/__manifest__.py]:
```suggestion
    "website": "https://github.com/OCA/web",
```

File: web_widget_heatmap/static/src/css/heatmap_widget.css [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/* Copyright 2019 Dinar Gabbasov <https://it-projects.info/team/GabbasovDinar>
```
Comment (Tardo) [inline web_widget_heatmap/static/src/css/heatmap_widget.css:1]: Please, use ~~'less'~~ 'Sass'

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        /**
+         * Find the records that match the selected date and show in the tree view
+         * @param {Date} date
+         * @returns {jQuery.Deferred} Action loaded
+         */
+        onClickHeatMap: function(date) {
+            date = moment(date);
+            var options = this.heatmap_options;
+            var elements = this.value ? _.pluck(this.value.data, 'data') : [];
+            if (elements && elements.length) {
```
Comment (Tardo) [inline web_widget_heatmap/static/src/js/heatmap_widget.js]:
```suggestion
            if (_.some(elements)) {
```

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                // TODO: make faster
+                var current_elements_ids = elements.filter(function(el) {
+                    var duration = moment.duration(date.diff(el.create_date));
+                    if (options.domain === 'hour') {
+                        return duration.years() === 0 && duration.months() === 0 && duration.weeks() === 0 && duration.days() === 0 && duration.hours() === 0;
```
Comment (Tardo) [inline web_widget_heatmap/static/src/js/heatmap_widget.js]: Can be better for not equal cases change the order:
```suggestion
                        return duration.hours() === 0 && duration.days() === 0 && duration.weeks() === 0 && duration.months() === 0 && duration.years() === 0;
```

**Review by Tardo — CHANGES_REQUESTED** [review]

File: web_widget_heatmap/static/src/css/heatmap_widget.css (hunk 5 above)
Comment (pedrobaeza) [reply web_widget_heatmap/static/src/css/heatmap_widget.css:1]: Better to use `sass` as `less` is deprecated

Comment (Tardo) [reply web_widget_heatmap/static/src/css/heatmap_widget.css:1]: Is 11.0

Comment (pedrobaeza) [reply web_widget_heatmap/static/src/css/heatmap_widget.css:1]: But sass is also present in 11.0. In 12.0, you have scss, with retro-compatibility with sass.

Comment (Tardo) [reply web_widget_heatmap/static/src/css/heatmap_widget.css:1]: Oh! thanks.

File: web_widget_heatmap/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        "web",
+    ],
+    "data": [
+        "views/web_widget_heatmap_template.xml",
+    ],
+    "demo": [],
+    "qweb": [
+        "static/src/xml/base.xml"
+    ]
+}
```
Comment (MeetKD) [inline web_widget_heatmap/__manifest__.py:30]: You can remove empty keys and may b change category to web?

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                        return duration.years() === 0 && duration.months() === 0;
+                    } else if (options.domain === 'year') {
+                        return duration.years() === 0;
+                    }
+                    return false;
+                }).map(function(el) {
+                    return el.id;
+                });
+                var action = {
+                    name: date.format('YYYY-MM-DD HH:mm'),
```
Comment (MeetKD) [inline web_widget_heatmap/static/src/js/heatmap_widget.js:178]: Can we take dateformat from l10n itself?

**Review by MeetKD — CHANGES_REQUESTED** [review]

File: web_widget_heatmap/static/src/css/heatmap_widget.css (hunk 3 above)
Comment (agent-z28) [reply web_widget_heatmap/static/src/css/heatmap_widget.css:4]: Thats 

> I like this widget.
> The loading problem can also be solved by making the widget work for Serialized fields. The programmer can make a computed and stored Serialized field which stores the records as JSON. This way every time the partner record opens, not all the sale orders are fetched instead just a text field essentially.

Thats the perfect solution, with a depends on the interesting fields

File: web_widget_heatmap/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    "development_status": "Stable",
+    "category": "Extra Tools",
+    "website": "https://github.com/OCA/web",
+    "author": "Dinar Gabbasov, Odoo Community Association (OCA)",
+    "maintainers": ["GabbasovDinar"],
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "preloadable": True,
+    "external_dependencies": {
```
Comment (daramousk) [inline web_widget_heatmap/__manifest__.py:16]: Remove mappings with no values.

File: web_widget_heatmap/static/lib/js/cal-heatmap.js [hunk 11]
```diff
@@ -0,0 +1,3487 @@
+/*! cal-heatmap v3.6.2 (Mon Oct 10 2016 01:36:20)
```
Comment (daramousk) [inline web_widget_heatmap/static/lib/js/cal-heatmap.js:1]: Use `cal-heatmap.min.js` instead.

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+*/
+
+odoo.define('web_widget_heatmap.widget', function (require) {
+    "use strict";
+
+    var AbstractField = require('web.AbstractField');
+    var registry = require('web.field_registry');
+
+
+    var HeatMapWidget = AbstractField.extend({
```
Comment (daramousk) [inline web_widget_heatmap/static/src/js/heatmap_widget.js:13]: If this widget can work with x2many fields, why not inherit from `FieldX2Many` instead?

File: web_widget_heatmap/static/src/js/heatmap_widget.js [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+        previousHeatMap: function() {
+            this.heatmap.previous();
+        },
+
+        /**
+         * Find the records that match the selected date and show in the tree view
+         * @param {Date} date
+         * @returns {jQuery.Deferred} Action loaded
+         */
+        onClickHeatMap: function(date) {
```
Comment (daramousk) [inline web_widget_heatmap/static/src/js/heatmap_widget.js:154]: I am not sure I understand this function, and the references to dates.

A heatmap's elements as I understand can represent a lot of things, not just dates/time. This heatmap could have as elements the records of the XMany field and depending on a field on those records proper opacity can be generated. 
I thin

**Review by daramousk — COMMENTED** [review]
Great contribution, thanks.

I was under the impression that something like this existed, and after looking a little bit around it looks that the module `web_graph` existed on 8.0 that did something like this, but it seems it has been removed, anyway I have left some comment in which I would like your opinion on.

Thanks

File: web_widget_heatmap/static/lib/js/cal-heatmap.js (hunk 11 above)
Comment (GabbasovDinar) [reply web_widget_heatmap/static/lib/js/cal-heatmap.js:1]: Why? Odoo automatically minimizes js files. This code format allows you to conveniently debug code in debug=assets mode

Comment (GabbasovDinar) [reply web_widget_heatmap/static/lib/js/cal-heatmap.js:1]: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#4javascript

Comment (legalsylvain) [reply web_widget_heatmap/static/lib/js/cal-heatmap.js:1]: OK to keep non minimizes files.

**Pull request closed without merging**